File: 500+/543-diameter-of-binary-tree.cpp
int diameterOfBinaryTree(TreeNode* root) {
    int diameter = 0;
    int tmp = helper(root, diameter);
    return diameter;
}

int helper(TreeNode* root, int &diameter) {
    if (root == nullptr) return 0;
    int left = helper(root->left, diameter), right = helper(root->right, diameter);
    diameter = max(left+right, diameter);
    return 1+max(left,right);
}

File: python version/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0

        def helper(node):
            if not node:
                return 0
            left_count = helper(node.left)
            right_count = helper(node.right)
            self.diameter = max(self.diameter, left_count+right_count)
            return 1 + max(left_count, right_count)    
        helper(root)
        return self.diameter